File: src/main/scala/network/LLMService.scala
package network

import java.net.URI
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.time.Duration
import ujson._

/**
 * Parsed interaction fields extracted by the LLM from natural language input.
 */
case class ParsedInteraction(
  personName: String,
  medium: String,
  location: String,
  theirLocation: Option[String],
  topics: List[String],
  note: Option[String],
  date: Option[String]
)

/**
 * Client for the OpenRouter API that extracts structured interaction data
 * from natural language descriptions.
 */
object LLMService {

  private val apiUrl = "https://openrouter.ai/api/v1/chat/completions"
  private val model = "meta-llama/llama-3.3-70b-instruct:free"

  /**
   * Reads the OpenRouter API key from the OPENROUTER_API_KEY environment variable.
   * Returns Left with an error message if not set.
   */
  def getApiKey: Either[String, String] = {
    val key = System.getenv("OPENROUTER_API_KEY")
    if (key == null || key.trim.isEmpty)
      Left("OPENROUTER_API_KEY environment variable is not set. Get a free key at https://openrouter.ai/keys")
    else
      Right(key.trim)
  }

  /**
   * Calls the LLM to extract structured interaction fields from a natural language description.
   *
   * @param input       The natural language description (e.g., "Had coffee with John at Starbucks")
   * @param knownNames  List of known contact names for accurate matching
   * @return Either an error message or the parsed interaction fields
   */
  def parseInteraction(input: String, knownNames: List[String], apiKey: String): Either[String, ParsedInteraction] = {
    val namesStr = knownNames.mkString(", ")
    val systemPrompt =
      s"""You extract interaction metadata from natural language descriptions.
         |Known contacts: [$namesStr]
         |Respond with JSON only, no other text.
         |JSON schema: { "personName": "...", "medium": "InPerson|Text|PhoneCall|VideoCall|SocialMedia", "location": "...", "theirLocation": null, "topics": ["..."], "note": null, "date": null }
         |Rules:
         |- personName must match one of the known contacts as closely as possible
         |- If the medium is not mentioned, default to "InPerson"
         |- location is where the interaction happened (required, use a reasonable default from context)
         |- theirLocation is only for remote interactions where their location differs; set to null for in-person
         |- topics should be a list of discussion subjects extracted from the description
         |- note is for any additional context not captured in other fields; set to null if none
         |- date should be "YYYY-MM-DD" if mentioned, otherwise null (today will be used)""".stripMargin

    val requestBody = ujson.Obj(
      "model" -> model,
      "response_format" -> ujson.Obj("type" -> "json_object"),
      "messages" -> ujson.Arr(
        ujson.Obj("role" -> "system", "content" -> systemPrompt),
        ujson.Obj("role" -> "user", "content" -> input)
      )
    )

    try {
      val client = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(15))
        .build()

      val request = HttpRequest.newBuilder()
        .uri(URI.create(apiUrl))
        .header("Content-Type", "application/json")
        .header("Authorization", s"Bearer $apiKey")
        .timeout(Duration.ofSeconds(30))
        .POST(HttpRequest.BodyPublishers.ofString(ujson.write(requestBody)))
        .build()

      val response = client.send(request, HttpResponse.BodyHandlers.ofString())

      if (response.statusCode() != 200) {
        return Left(s"API request failed (HTTP ${response.statusCode()}): ${response.body().take(200)}")
      }

      val json = ujson.read(response.body())
      val content = json("choices")(0)("message")("content").str
      parseJsonResponse(content)
    } catch {
      case e: java.net.http.HttpTimeoutException =>
        Left("Request timed out. Please try again.")
      case e: java.net.ConnectException =>
        Left("Could not connect to OpenRouter API. Check your internet connection.")
      case e: Exception =>
        Left(s"Unexpected error: ${e.getMessage}")
    }
  }

  /**
   * Parses the JSON string returned by the LLM into a ParsedInteraction.
   */
  private def parseJsonResponse(content: String): Either[String, ParsedInteraction] = {
    try {
      val json = ujson.read(content)
      val personName = json("personName").str
      val medium = json("medium").str
      val location = json("location").str
      val theirLocation = json.obj.get("theirLocation").flatMap(v => if (v.isNull) None else Some(v.str))
      val topics = json("topics").arr.map(_.str).toList
      val note = json.obj.get("note").flatMap(v => if (v.isNull) None else Some(v.str))
      val date = json.obj.get("date").flatMap(v => if (v.isNull) None else Some(v.str))

      if (personName.isEmpty) Left("LLM did not extract a person name")
      else if (topics.isEmpty) Left("LLM did not extract any topics")
      else Right(ParsedInteraction(personName, medium, location, theirLocation, topics, note, date))
    } catch {
      case e: Exception =>
        Left(s"Failed to parse LLM response: ${e.getMessage}")
    }
  }
}

File: src/ai/llm_service.rs
use std::env;
use serde::{Deserialize, Serialize};

const DEFAULT_OLLAMA_URL: &str = "http://localhost:11434";
const DEFAULT_MODEL: &str = "llama3.2:3b";

/// A past correction: what the AI parsed vs what the user actually saved.
pub struct CorrectionExample {
    pub original_text: String,
    pub ai_output: String,
    pub user_output: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParsedInteraction {
    #[serde(default, alias = "personName")]
    pub person_names: Vec<String>,
    #[serde(default = "default_medium")]
    pub medium: String,
    #[serde(default)]
    pub location: String,
    pub their_location: Option<String>,
    #[serde(default)]
    pub topics: Vec<String>,
    pub note: Option<String>,
    pub date: Option<String>,
}

fn default_medium() -> String {
    "InPerson".into()
}

fn ollama_url() -> String {
    env::var("OLLAMA_HOST").unwrap_or_else(|_| DEFAULT_OLLAMA_URL.to_string())
}

fn ollama_model() -> String {
    env::var("PRM_MODEL").unwrap_or_else(|_| DEFAULT_MODEL.to_string())
}

/// Check that Ollama is running and reachable.
pub fn check_ollama() -> Result<(), String> {
    let url = ollama_url();
    ureq::get(&url)
        .timeout(std::time::Duration::from_secs(3))
        .call()
        .map_err(|_| format!(
            "Cannot connect to Ollama at {}. Is it running?\n  Install: https://ollama.ai\n  Start:   ollama serve",
            url
        ))?;
    Ok(())
}

pub fn parse_interaction(
    input: &str,
    known_names: &[String],
    corrections: &[CorrectionExample],
) -> Result<ParsedInteraction, String> {
    let names_str = known_names.join(", ");
    let today = chrono::Local::now().format("%Y-%m-%d");
    let model = ollama_model();

    let corrections_block = if corrections.is_empty() {
        String::new()
    } else {
        let mut block = String::from("\nPast corrections to learn from (most recent first):\n");
        for (i, c) in corrections.iter().enumerate() {
            block.push_str(&format!(
                "\nExample {}:\nInput: {}\nYou parsed: {}\nUser corrected to: {}\n",
                i + 1,
                c.original_text,
                c.ai_output,
                c.user_output,
            ));
        }
        block.push_str("\nApply these learnings when parsing the new input.\n");
        block
    };

    let system_prompt = format!(
        r#"You extract interaction metadata from natural language descriptions.
Today's date is {today}.
Known contacts: [{names_str}]
Respond with JSON only, no other text.
JSON schema: {{ "personNames": ["..."], "medium": "InPerson|Text|PhoneCall|VideoCall|SocialMedia", "location": "...", "theirLocation": null, "topics": ["..."], "note": null, "date": null }}
Rules:
- personNames is an array of ALL people mentioned in the input. If multiple people are mentioned, include all of them. Use names exactly as written. Match to known contacts when possible. NEVER substitute different names.
- If the medium is not mentioned, default to "InPerson"
- location: use the most specific location from the input (e.g. "Charlton, MA" not just "home"). Include the full place name, city, or address as given.
- theirLocation is only for remote interactions where their location differs; set to null for in-person
- topics: ONLY include activities or subjects explicitly mentioned in the input. Do NOT infer or add topics that weren't stated. Be aware of slang (e.g. "gas" means great/amazing, not cooking).
- note is for any additional context not captured in other fields; set to null if none
- date: ONLY set to a "YYYY-MM-DD" string if the user explicitly mentions a specific date (e.g. "yesterday", "last Friday", "on March 5th"). Otherwise MUST be null. null means today.{corrections_block}"#
    );

    let request_body = serde_json::json!({
        "model": model,
        "messages": [
            { "role": "system", "content": system_prompt },
            { "role": "user", "content": input }
        ],
        "format": "json",
        "stream": false
    });

    let api_url = format!("{}/api/chat", ollama_url());

    let response = ureq::post(&api_url)
        .set("Content-Type", "application/json")
        .timeout(std::time::Duration::from_secs(120))
        .send_json(request_body)
        .map_err(|e| match e {
            ureq::Error::Status(code, resp) => {
                let body = resp.into_string().unwrap_or_default();
                format!("Ollama request failed (HTTP {}): {}", code, &body[..body.len().min(200)])
            }
            ureq::Error::Transport(t) => {
                if t.to_string().contains("timed out") {
                    "Request timed out. Local models can be slow on first run — try again.".into()
                } else {
                    format!("Could not connect to Ollama: {}", t)
                }
            }
        })?;

    let json: serde_json::Value = response
        .into_json()
        .map_err(|e| format!("Failed to parse Ollama response: {}", e))?;

    let content = json
        .pointer("/message/content")
        .and_then(|v| v.as_str())
        .ok_or("No content in Ollama response")?;

    parse_llm_json(content)
}

// Custom deserialization from the LLM JSON which uses camelCase
fn parse_llm_json(content: &str) -> Result<ParsedInteraction, String> {
    let json: serde_json::Value =
        serde_json::from_str(content).map_err(|e| format!("Failed to parse LLM response: {}", e))?;

    let person_names = match json.get("personNames") {
        Some(v) if v.is_array() => v
            .as_array()
            .unwrap()
            .iter()
            .filter_map(|v| v.as_str().map(|s| s.to_string()))
            .filter(|s| !s.is_empty())
            .collect(),
        _ => {
            // Fallback to single personName
            match json.get("personName").and_then(|v| v.as_str()) {
                Some(name) if !name.is_empty() => vec![name.to_string()],
                _ => Vec::new(),
            }
        }
    };

    let medium = json
        .get("medium")
        .and_then(|v| v.as_str())
        .unwrap_or("InPerson")
        .to_string();

    let location = json
        .get("location")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let their_location = json
        .get("theirLocation")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string());

    let topics: Vec<String> = json
        .get("topics")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(|s| s.to_string()))
                .collect()
        })
        .unwrap_or_default();

    let note = json
        .get("note")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string());

    let date = json
        .get("date")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string());

    if person_names.is_empty() {
        return Err("LLM did not extract any person names".into());
    }
    if topics.is_empty() {
        return Err("LLM did not extract any topics".into());
    }

    Ok(ParsedInteraction {
        person_names,
        medium,
        location,
        their_location,
        topics,
        note,
        date,
    })
}
